### PR TITLE
[mxml] New Port

### LINKED ISSUE #19042: [New Port Request] Mini-XML
Library name: Mini-XML

Library description: Mini-XML is a tiny XML library that you can use to read and write XML and XML-like data files in your application without requiring large non-standard libraries.

Source repository URL: https://github.com/michaelrsweet/mxml

Project homepage (if different from the source repository): https://www.msweet.org/mxml/

Anything else that is useful to know when adding (such as optional features the library may have that should be included):

### PR DESCRIPTION
**Pull request by Rossmaxx**
fixes: #19042 

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by Rossmaxx** [comment]
Do note that this PR is incomplete. I don't really know what I'm doing here, so any docs and extra help and info appreciated. Most of what's written here is copied from other ports.

**Comment by Rossmaxx** [comment]
What's with the license fail?

**Comment by FrankXie05** [comment]
@Rossmaxx Thanks for this PR. I think we need to ask upstream whether to add it to vcpkg. In this way, I can help you complete this PR. :)

**Comment by Rossmaxx** [comment]
> I think we need to ask upstream whether to add it to vcpkg

I did ask about this upstream some time ago, and here's what the maintainer said. Forgot that i asked this in the first place.

https://github.com/michaelrsweet/mxml/pull/281#issuecomment-1763439329

I initially asked for cmake support but the maintainer refused. It's after some time that I understand that vcpkg doesn't require ports to be in cmake and other build systems being supported.

**Comment by dg0yt** [comment]
(I don't understand the comments about upstream. No patches, no restriction, no need to ask.)

File: ports/mxml/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "name": "mxml",
+    "version": "3.3.1",
+    "homepage": "https://github.com/michaelrsweet/mxml",
+    "description": "A tiny xml parsing library",
+    "license": "Apache license 2.0"
```
[the remark this answers is not in the document]
Comment (Rossmaxx) [reply ports/mxml/vcpkg.json]: Didn't see this for all this time.

**Comment by Rossmaxx** [comment]
Quite a lot has happened since the last time, Upstream has moved on to version 4, while this port is at version 3. For convenience sake, I'll be working on version 3, and the upstream suggests mxml4 to be a seperate package because of a breakage in compatibility. 

@FrankXie05 @dg0yt sorry for the delay. Can you guys help me move this PR forward?

**Comment by dg0yt** [comment]
> the upstream suggests mxml4 to be a seperate package because of a breakage in compatibility.

AFAICT vcpkg is reluctant to accept a "separate package". Usually it is not possible to install two variants of a package at the same time, but this is what vcpkg CI does.

**Comment by Rossmaxx** [comment]
For my use case, mxml 3 is fine. If need arises, I can make a seperate PR to bump the version.

**Comment by dg0yt** [comment]
For !windows, try at least with `DETERMINE_BUILD_TRIPLET` instead of `AUTOCONFIG`.

**Comment by Rossmaxx** [comment]
Everything mostly seems to work now, except for a few loose ends.
```log
Building mxml[core]:x64-windows...
-- Using cached michaelrsweet-mxml-fd47c7d115191c8a6bce2c781ffee41e179530f2.tar.gz.
-- Cleaning sources at C:/vcpkg/buildtrees/mxml/src/1e179530f2-6bf12190f1.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/vcpkg/downloads/michaelrsweet-mxml-fd47c7d115191c8a6bce2c781ffee41e179530f2.tar.gz
-- Using source at C:/vcpkg/buildtrees/mxml/src/1e179530f2-6bf12190f1.clean
-- Getting CMake variables for x64-windows
-- Building vcnet/mxml1.vcxproj for Release
-- Building vcnet/mxml1.vcxproj for Debug
-- Installing: C:/vcpkg/packages/mxml_x64-windows/share/mxml/copyright
-- Performing post-build validation
warning: The folder /include is empty or not present. This indicates the library was not correctly installed.
error: Found 1 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile: C:\vcpkg\ports\mxml\portfile.cmake
Stored binary cache: "C:\Users\ROSHAN\AppData\Local\vcpkg\archives\98\98f5f42675b912e3f5b0e565833259e9b2ec105dce331332f03ec13c985639e2.zip"
Elapsed time to handle mxml:x64-windows: 22.42 s
```

can anyone help?

> For !windows, try at least with DETERMINE_BUILD_TRIPLET instead of AUTOCONFIG.

can you explain?

**Comment by dg0yt** [comment]
`vcpkg_configure_make` was called with `AUTOCONFIG`. This is an extra step which failed. But it is optional. `DETERMINE_BUILD_TRIPLET` is enough to ensure proper configuration also for cross builds.

**Comment by Rossmaxx** [comment]
Didn't I restrict builds for windows?

**Comment by Rossmaxx** [comment]
I need help with post build validation, @dg0yt @FrankXie05

**Comment by FrankXie05** [comment]
># there's something wrong here
warning: The folder /include is empty or not present. This indicates the library was not correctly installed.
error: Found 1 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile: C:\vcpkg\ports\mxml\portfile.cmake
Stored binary cache: 

This port build does not install any header files available,
Add `set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)` to the top of `portfile.cmake` or install the header files manually through `file (INSTALL )`

**Comment by dg0yt** [comment]
> Add `set(VCPKG_POLICY_EMPTY_INCLUDE_FOLDER enabled)` to the top of `portfile.cmake` 

Seriously?

msbuild doesn't install headers, so it must be done by the portfile.
The autotools build would tell us which destinations are desired, but it is omitted.
Which also mean I cannot test and help.

**Comment by JohannesLorenz** [comment]
I would rather favor mxml4, because there will be a point of time where mxml3 will be outdated on all Linux systems, and in that case, we need a second vcpkg port.

Given all the issues with autotools, would it be possible if we just maintain a CMake branch/fork and just install that? Of course, I would also prefer to make it work with autotools.

**Comment by dg0yt** [comment]
It is unlikely that vcpkg will accept both mxml3 and mxml4.
It is unlikely that vcpkg will accept a fork.
And I'm not convinced there is a serious problem with autotools.

**Comment by Rossmaxx** [comment]
> I would rather favor mxml4,

You need to port zyn code to work on mxml4 in that case. I can change it quick. 

> I'm not convinced there is a serious problem with autotools.

I am not saying there's a problem with auto tools, I'm just unfamiliar with pkg-config and also, how am i supposed to integrate an autotools build with cmake projects?

**Comment by dg0yt** [comment]
> I'm just unfamiliar with pkg-config and also, how am i supposed to integrate an autotools build with cmake projects?

ATM this PR provides neither CMake integation nor pkg-config, and AFAIU there is no official CMake support from upstream. And you don't need to build with CMake to add an unofficial config module - it is just a CMake file with the right name, location, and code.

And see e.g. `scripts/test_ports/libavif` how to write a small test that pkg-config and cmake do the right thing.

**Comment by JohannesLorenz** [comment]
> You need to port zyn code to work on mxml4 in that case. I can change it quick.

Done in zynaddsubfx/zynaddsubfx#315 and merged last week :smiley:

**Comment by Rossmaxx** [comment]
All righty, I'll switch it up to use mxml 4 completely

**Comment by Rossmaxx** [comment]
I have upgraded this port to mxml v4, and it's better to leave v3 behind. I am also trying to add support for non-windows builds now, but the CI is failing. I'll try troubleshooting from my linux partition later as i'm done with this for now. Did I make a mistake in the else part, or did I miss something? 

I did fix the empty include folder issue tho, and I think it's coming along nicely.

> vcpkg_configure_make was called with AUTOCONFIG. This is an extra step which failed. But it is optional. DETERMINE_BUILD_TRIPLET is enough to ensure proper configuration also for cross builds.

I remember seeing this before, but I forgot. Seeing this again now after putting the work into it.

I now also need to figure out a way to make this port work on cmake projects, if it doesn't by default.

@dg0yt more help please.

**Comment by dg0yt** [comment]
x64-linux CI:
~~~
make: *** No rule to make target 'mxml-attr.c', needed by 'mxml-attr.o'.  Stop.
~~~
Strange, but it might have an easy explanation. I could check later.

**Comment by Rossmaxx** [comment]
Is this now an issue with the upstream's makefile?

**Comment by Rossmaxx** [comment]
here's the upstream, if it matters

**Review by dg0yt — CHANGES_REQUESTED** [review]
For non-Windows, the key is `DETERMINE_BUILD_TRIPLET` and `COPY_SOURCE`.

File: ports/mxml/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,26 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO michaelrsweet/mxml
+    REF "v${VERSION}"
+    SHA512 3a90d9929b10fb563a8b5dcde67ec766e306397211c1125954064f7ca207d0d246eaf839f9fbf36e049aaab370670b97debc689121ea9f1c9c2edf0ea3a32eb9
```
[the remark this answers is not in the document]
Comment (Rossmaxx) [reply ports/mxml/portfile.cmake:5]: Wait, when did that change? After I changed the tag?

File: ports/mxml/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        TARGET Build
+    )
+ELSE()
+    vcpkg_configure_make(
+        SOURCE_PATH "${SOURCE_PATH}"
+        AUTOCONFIG
+    )
+    vcpkg_install_make()
+    vcpkg_fixup_pkgconfig()
+ENDIF()
```
Comment (dg0yt) [inline ports/mxml/portfile.cmake:23]:
```suggestion
endif()

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```

File: ports/mxml/portfile.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (dg0yt) [reply ports/mxml/portfile.cmake:5]: When the tag changes, the SHA512 will need to change, too. Asset caching as in CI will hide the problem because there is a cache hit for the outdated SHA512.

File: ports/mxml/portfile.cmake (hunk 3 above)
Comment (Rossmaxx) [reply ports/mxml/portfile.cmake:25]: Why this?

**Comment by Rossmaxx** [comment]
I'm doing all this commits from my phone, I'll take some time to fix the versions.

**Comment by Rossmaxx** [comment]
What caused this build fail now? I'll investigate later.

**Comment by Rossmaxx** [comment]
Well, this is strange, it builds fine locally. What's with the CI fail then?

**Comment by dg0yt** [comment]
Locally, post-build checks are not fatal. For the official registry, they are.
~~~
-- Performing post-build validation
D:\a\_work\1\s\ports\mxml\portfile.cmake: warning: DLLs should not be present in a static build, but the following DLLs were found. To suppress this message, add set(VCPKG_POLICY_DLLS_IN_STATIC_LIBRARY enabled)
D:\p\mxml_x64-windows-static: note: the DLLs are relative to ${CURRENT_PACKAGES_DIR} here
note: debug/bin/mxml4.dll
note: bin/mxml4.dll
D:\a\_work\1\s\ports\mxml\portfile.cmake: warning: ${CURRENT_PACKAGES_DIR}/debug/bin exists but should not in a static build. To suppress this message, add set(VCPKG_POLICY_DLLS_IN_STATIC_LIBRARY enabled)
D:\a\_work\1\s\ports\mxml\portfile.cmake: warning: ${CURRENT_PACKAGES_DIR}/bin exists but should not in a static build. To suppress this message, add set(VCPKG_POLICY_DLLS_IN_STATIC_LIBRARY enabled)
note: if creation of these directories cannot be disabled, you can add the following in portfile.cmake to remove them
if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
  file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/bin" "${CURRENT_PACKAGES_DIR}/bin")
endif()
D:\a\_work\1\s\ports\mxml\portfile.cmake: warning: Found 3 post-build check problem(s). These are usually caused by bugs in portfile.cmake or the upstream build system. Please correct these before submitting this port to the curated registry.
error: building mxml:x64-windows-static failed with: POST_BUILD_CHECKS_FAILED
See https://learn.microsoft.com/vcpkg/troubleshoot/build-failures?WT.mc_id=vcpkg_inproduct_cli for more information.
~~~
So basically, it always builds DLLs, but this unexpected for static triplets.
- You may patch the msbuild project to create static libs.
- You may make explicit the limitation with:
  ~~~cmake
  if(VCPKG_TARGET_IS_WINDOWS)
      vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
  endif()
  ~~~
  but this implies `"supports": "!(windows & static & staticcrt)`.
- You may try to use the autotools build also for MSVC. But I have no idea if it works and helps for this project.

**Comment by Rossmaxx** [comment]
Ahh, forgot that static builds are a thing. 

I should maybe patch in support for static builds as I'm pretty sure mxml is used as a static library in zyn. @JohannesLorenz is it so?

**Comment by JohannesLorenz** [comment]
> I should maybe patch in support for static builds as I'm pretty sure mxml is used as a static library in zyn. @JohannesLorenz is it so?

@Rossmaxx When we compile, we just say link against it - it does not matter. For example, on my Linux, mxml is a shared library (sorry for the late reply!)

**Comment by Rossmaxx** [comment]
I seriously don't know what to do with this PR now. This is one of those PRs which I silently abandoned due to the passage of time. I saw old mails and saw this just now. 

I can't say when I'll get back to this either. If any of the other participants have any more idea than me, please drop them. Or if anyone plans to take over, that's welcome too.

**Comment by Rossmaxx** [comment]
I have disabled static builds so that it can be merged soon. If anyone wants to have static library, they can continue working in a seperate pull req. Get this one merged soon pls.

**Comment by Rossmaxx** [comment]
Why is the mac os arm64 workload still running, though i clearly disabled it in vcpkg.json
@dg0yt ?

Also this PR is ready on my end

**Comment by Rossmaxx** [comment]
I Just noticed that 32 bit builds are failing too, I'll look tmrw. If it's something I can't fix, I'll disable that too for now. 

Also, another developer has offered to take up this effort which @JohannesLorenz informed me on discord, and once this PR lands on master, they can follow up.

**Comment by Rossmaxx** [comment]
Apparently the issue is just an outdated branch. Rebasing on master fixed it.

**Comment by Rossmaxx** [comment]
Now i need to fix up support and refine it.

**Comment by Rossmaxx** [comment]
@LilyWangLL The PR is ready

**Comment by Rossmaxx** [comment]
Bump

File: ports/mxml/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
+    vcpkg_msbuild_install(
+        SOURCE_PATH "${SOURCE_PATH}"
+        PROJECT_SUBPATH "vcnet/mxml4.vcxproj"
+        TARGET Build
+    )
+    file(INSTALL "${SOURCE_PATH}/mxml.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+else()
+    vcpkg_configure_make(
```
Comment (BillyONeal) [inline ports/mxml/portfile.cmake]: Can you use the `vcpkg_make_Xxx` family instead?

Comment (Rossmaxx) [reply ports/mxml/portfile.cmake]: done

**Comment by Rossmaxx** [comment]
@BillyONeal Good for merge now?

**Review by BillyONeal — APPROVED** [review]
Thanks for the new port!
